src/container/TreeContainer/Base/index.ts

```diff
@@ -1,20 +1,82 @@
-import TreeNode, { TreeNodeColor } from './TreeNode';
-import TreeIterator from './TreeIterator';
-import { Container } from '@/container/ContainerBase/index';
+import type TreeIterator from './TreeIterator';
+import { Container } from '@/container/ContainerBase';
 import { $checkWithinAccessParams } from '@/utils/checkParams.macro';
+import { TreeNode, TreeNodeColor, TreeNodeEnableIndex } from './TreeNode';
 
 abstract class TreeContainer<K, V> extends Container<K | [K, V]> {
   protected root: TreeNode<K, V> | undefined = undefined;
-  protected header: TreeNode<K, V> = new TreeNode<K, V>();
+  protected header: TreeNode<K, V>;
   protected cmp: (x: K, y: K) => number;
-  protected constructor(cmp: (x: K, y: K) => number =
-  (x: K, y: K) => {
-    if (x < y) return -1;
-    if (x > y) return 1;
-    return 0;
-  }) {
+  TreeNodeClass: typeof TreeNode | typeof TreeNodeEnableIndex;
+  /**
+   * @description Whether node indexing is enabled.
+   */
+  private enableIndex: boolean;
+  /**
+   * @description Remove a node.
+   * @param curNode The node you want to remove.
+   * @protected
+   */
+  protected eraseNode: (curNode: TreeNode<K, V>) => void;
+  /**
+   * @description Insert a key-value pair or set value by the given key.
+   * @param key The key want to insert.
+   * @param value The value want to set.
+   * @param hint You can give an iterator hint to improve insertion efficiency.
+   * @protected
+   */
+  protected set: (key: K, value: V, hint?: TreeIterator<K, V>) => void;
+  protected constructor(
+    cmp: (x: K, y: K) => number =
+    (x: K, y: K) => {
+      if (x < y) return -1;
+      if (x > y) return 1;
+      return 0;
+    },
+    enableIndex = false
+  ) {
     super();
     this.cmp = cmp;
+    this.enableIndex = enableIndex;
+    if (enableIndex) {
+      this.TreeNodeClass = TreeNodeEnableIndex;
```

Comment:
This kind of code injection does not become tree shaking
In order to properly implement this, it is necessary to implement an object with code to be handed over from the constructor.

The mechanism of tree-shaking is to create a chain of dependency graphs for what the user imports and what the imported file imports again.
So we can induce treeshaking by having the user directly import or not import certain code.

for example:
``` typescript
//ContainerFeature.ts
interface ContainerFeature {
  eraseNode(): void;
}
```
``` typescript
//ContainerFeature1.ts
import type { ContainerFeature } from './ContainerFeature';

export class ContainerFeature1 implements ContainerFeature {
  eraseNode() {
    //override implements    
  }
}
```
``` typescript
//Container.ts
import type { ContainerFeature } from './ContainerFeature';
// The point is that the ContainerFeature1 type is not imported from here.

export class Container {
  constructor(optionalFeature?: ContainerFeature) {
    if (optionalFeature) {
      this.eraseNode = optionalFeature.eraseNode; // inject feature
    }
  }

  eraseNode() {
    //original implements    
  }
}
```

Reply:
But the subclass will import the ContainerFeature1, so tree shaking will also fail?

Reply:
I don't know what's the point.

If I understand correctly, it will be:

```typescript
// Base/index.ts
abstract class TreeContainer {
  constructor(optionalFeature?: typeof ContainerFeature1 | typeof ContainerFeature2) {
    this.eraseNode = optionalFeature.eraseNode;
  }
}
export default TreeContainer;
```

```typescript
// OrderedSet.jts
import Tree
export abstract class TreeContainer {
  constructor(enableIndex: boolean) {
    // ${1}
  }
}
```

I don't know how to implement `${1}` because the `super` must in the first line of constructor.

Did you mean let user give the `optionalFeature` param? If yes, I think it's a bad operation.

Reply:
Right, you can expose the optionalFeature to the user and show it as if it is enum.

Reply:
From the user's point of view, it seems as if the enableIndex attribute has changed to enum. But in fact, it's not an enum, it's an object that includes an implementation.

Reply:
Is that nesserary?

This change will make `TreeContainer` difficult to use.

Reply:
The way you use it doesn't change at all, it's just that you have to put an implementation when you use a new feature

``` typescript
// normal ordered set that doesn't have iterator index feature
const set = new OrderedSet<number>();

// ordered set that does have iterator index feature
const indexSet = new OrderedSet<number>(undefined, undefined, UseIteratorIndex);
```

Certainly, putting objects directly in order to use the Iterator Index feature is a cumbersome task. But there's nothing we can do to maintain the code size

Reply:
We could optimize when this part of the code size is large, but now I don't think it's necessary.

We can do it in the future.